lib/SILOptimizer/Mandatory/DiagnoseVarUsage.cpp

```diff
@@ -1084,6 +1116,16 @@ class VariableReadTraverser {
         }
       }
     }
+    
+    // As a last resort, check if the value was discarded in an
+    // underscore assignment.
+    if (auto AFD = dyn_cast<AbstractFunctionDecl>(getVarDecl(EntryInst)->
```

Comment:
Walking the AST like this is a layering violation to me. At the very least, this won't handle closure bodies or other non-AbstractFunctionDecl local contexts.

Reply:
The reason `_ = foo` produces no code is because SILGen peepholes it away. Perhaps we should just not do that, and allow dead code elimination to remove the discarded result after the diagnostics pass?

Reply:
> Perhaps we should just not do that, and allow dead code elimination to remove the discarded result after the diagnostics pass

This definitely sounds like the ideal solution here. I will look into this. Thanks!

Reply:
This has been addressed in my latest commits! The peephole has been removed, and this makes uses of a `var` visible.. Even still, `let` constants do not have any instructions to generate, so I created a non-canonical `mark_discarded` instruction for cases like this.